### PR TITLE
Fix #1251 -- Event list/calendar: Show "event almost sold out" state

### LINKED ISSUE #1251: Show ticket availability by colour system
I would like to provide my customers the possibility to see if there are still lots of tickets availably, or if it is already nearly sold out. I know that there is the possibility to show the number of available tickets, but this information I don't want to show in such detail.

I think it would be nice to have a traffic light system. If more than 60% tickets are available, display a green icon, between 60 and 100 a yellow one, and if sold out a red one.

### PR DESCRIPTION
**Pull request by raphaelm**
This introduces a new setting as well as a new state in the event list and calendar. If set the new setting to "20", the event calendar will now mark all tickets with less than 20% available spots as "Few tickets left".

2023-01-26-234606_1381x459_scrot

This is a little complex to implement as we need to deal with all possible combinations of overlapping quotas, etc.
The `best_availability_state` function is entirely rewritten for this purpose. In the end, performance should be approximately the same as before.

Design-wise, I think there are two main open questions:

- Should this new state be green or yellow? I kept it green for now since we use yellow for "waiting list", "reserved", and "not yet for sale" currently. So yellow currently means "not able to book", and making this new state yellow could be confusing. On the other hand, making the new state yellow will make it more visible, and is also consistent with the traffic light analogy used on many other sites.
- If the setting is enabled, should we also show something on the event detail page? Should it be a "low on tickets" above the list of events, or should we show a similar yellow indicator for every low-inventory item? That would be more useful but also "leak" more info.
- Is the wording good like this? We tend to avoid "ticket" in this place, but in this case I think it makes sense (since the feature can be disabled) and I did not want something longer.

We should also add more edge cases with weird quotas to the tests, I guess.

### CODE REVIEW

**Comment by wiffbi** [comment]
Generally I think the tricky part is deciding when the event as a whole is low on inventory – if it is just tickets, that might be easy. But what if it is limited on-site-venue-tickets, but unlimited streaming-tickets or exhibition-tickets? Could a situation happen, where we do not show the overall „almost sold out“ but in fact, the one most wanted product is already sold out (sorry, I have not yet looked at the implementation of `best_availability_state`). Or even the other way around, we show the „almost sold out“ due to some side products running low, but the main product being available and this maybe causing lots of support requests to the organizer of why the event shows as almost sold out. In that case, they probably just need disable that setting as it is not that useful.

> * Should this new state be green or yellow? I kept it green for now since we use yellow for "waiting list", "reserved", and "not yet for sale" currently. So yellow currently means "not able to book", and making this new state yellow could be confusing. On the other hand, making the new state yellow will make it more visible, and is also consistent with the traffic light analogy used on many other sites.

I looked on the calendar-pages and `.waitinglist` does not have any special treating (although there is a CSS-class for it) and uses the default `$brand-primary`. `.reserved` uses `$brand-danger`. `$brand-warning` is not used at all. First of all, I think this should be harmonized and colors should be used the same for state on calender-view as well as list view. I also think, `$brand-warning` (or in our default case yellow) signals „almost sold out“ best and is a common pattern. The only problem will be accessiblity due to low color contrast with white text on yellow background. We would need to change the status-label to e.g. something like on the order-page, where order-status is actually just a colored dot with black on white text next to it. Hm …

> * If the setting is enabled, should we also show something on the event detail page? Should it be a "low on tickets" above the list of events, or should we show a similar yellow indicator for every low-inventory item? That would be more useful but also "leak" more info.

That is a good question. I see both as valid options and the first impulse would be „make it a setting“. But then, hm. I think the overall „almost sold out“ warning should be there in any case, just to make it consistent with the event list. What about we start with the overall warning/status and maybe later add a per item inventory if needed?

> * Is the wording good like this? We tend to avoid "ticket" in this place, but in this case I think it makes sense (since the feature can be disabled) and I did not want something longer.

I think it works, but you are right, we should try to avoid „ticket“. What about „Almost sold out“?

**Comment by raphaelm** [comment]
> Generally I think the tricky part is deciding when the event as a whole is low on inventory – if it is just tickets, that might be easy. But what if it is limited on-site-venue-tickets, but unlimited streaming-tickets or exhibition-tickets? Could a situation happen, where we do not show the overall „almost sold out“ but in fact, the one most wanted product is already sold out (sorry, I have not yet looked at the implementation of best_availability_state). Or even the other way around, we show the „almost sold out“ due to some side products running low, but the main product being available and this maybe causing lots of support requests to the organizer of why the event shows as almost sold out. In that case, they probably just need disable that setting as it is not that useful.

Yeah, of course this can be a problem with all the complexity pretix supports, but we already have two fixes for it:

- We have the `Quota.ignore_for_event_availability` config option (i.e. when you have t-shirts that you don't want to be counted for the events availability)
- The new feature is opt-in, if it doesn't fit your use case well, don't use it

I don't think we should hold off on this much-requested features just because it cannot solve every edge case.

> I looked on the calendar-pages and .waitinglist does not have any special treating (although there is a CSS-class for it) and uses the default $brand-primary.  .reserved uses $brand-danger. $brand-warning is not used at all.

Interesting, we are apparently inconsistent there. Why are we inconsistent there? We use $brand-warning for both waiting list as well as reserved at least in the list form of the calendar and in the widget!

If we change the widget though, it's a breaking change some users might not like (but they can override with custom CSS, so it's not that bad?)

> First of all, I think this should be harmonized and colors should be used the same for state on calender-view as well as list view. I also think, $brand-warning (or in our default case yellow) signals „almost sold out“ best and is a common pattern. The only problem will be accessiblity due to low color contrast with white text on yellow background.

Agreed. 

> We would need to change the status-label to e.g. something like on the order-page, where order-status is actually just a colored dot with black on white text next to it. Hm …

I don't really like it, but I guess we need to :-( 

> That is a good question. I see both as valid options and the first impulse would be „make it a setting“.

Not even more settings please!

> But then, hm. I think the overall „almost sold out“ warning should be there in any case, just to make it consistent with the event list. What about we start with the overall warning/status and maybe later add a per item inventory if needed?

We already have the option to show exact inventory per item. I don't want another per-item option. So +1

> I think it works, but you are right, we should try to avoid „ticket“. What about „Almost sold out“?

Hmm. I'd prefer a positive wording? "Almost sold out" sounds more like the negative.

**Comment by wiffbi** [comment]
> > We would need to change the status-label to e.g. something like on the order-page, where order-status is actually just a colored dot with black on white text next to it. Hm …
> 
> I don't really like it, but I guess we need to :-(

What about a combined yellow-green label? Base-color would be green, but with an added yellow. I did a quick „boxy“ test – could be a gradient as well, but I think we do not use gradients anywhere so it would feel a bit „alien“. Benefit of using this combined label, we could keep other label colors as-is (and harmonize $brand-warning on calendar-view). It sill communicates „available“ although with bit of a warning. I like the second one with yellow on the left better. That type of design could maybe be used in calendar view as well with the base color being green, but that strong color bar being yellow. Could work, could fail, but we might try.

Bildschirm­foto 2023-01-27 um 10 50 45
Bildschirm­foto 2023-01-27 um 10 49 27

> 
> > I think it works, but you are right, we should try to avoid „ticket“. What about „Almost sold out“?
> 
> Hmm. I'd prefer a positive wording? "Almost sold out" sounds more like the negative.

You are right, then probably your wording is better. I will think about it some more.

**Comment by raphaelm** [comment]
I like your idea. However, putting just an empty border to the left of the label looked off to me, since it broke the alignment with other labels in the lines above and below. So I put an icon into the border and I think it's nice now?



Similar to your suggestion, I changed just the border color in the calendar view. It's not perfect, but it works:

2023-01-30T18:47:04+01:00

For the widget, calendar entries currently don't have a border, so I added one only in this case to maintain compatibility, but on the right side to keep alignment of texts:



In the widget list mode, I put it to the left, again for keeping alignment of texts:

2023-01-30T18:47:34+01:00

I also made "reserved" consistently red (like in the calendar previously, I think it makes most sense) and waiting list consistently yellow (like in the list previously).

**Review by wiffbi — COMMENTED** [review]
Looks good, just some minor layout comments. I thought about adding an early return `if best_state_found == QUota.AVAILABILITY_OK`, but of so, we cannot calculate `num_tickets_found` and `num_tickets_possible` correctly.

File: src/pretix/static/pretixpresale/scss/_theme.scss [hunk 1]
```diff
@@ -121,3 +121,30 @@ footer {
...
   }
 }
+
+.label-success-warning {
+  @include label-variant($label-success-bg);
+
+  padding-left: 2.5em;
+  position: relative;
+  &::before {
+    font-family: FontAwesome;
```
[the remark this answers is not in the document]
Comment (raphaelm) [reply src/pretix/static/pretixpresale/scss/_theme.scss:131]: Yeah, just the exclamation mark is better!

File: src/pretix/base/models/event.py [hunk 2]
```diff
@@ -356,7 +356,7 @@ def best_availability_is_low(self):
         if ba[1] is None or not ba[2]:
             return False
 
-        percentage = ba[1] / ba[2]
+        percentage = ba[1] / ba[2] * 100
```
Comment (wiffbi) [inline src/pretix/base/models/event.py:359]: Good catch!
